libyul/AST.h

```diff
@@ -36,38 +39,85 @@ namespace solidity::yul
 
 using Type = YulString;
 
-struct DebugData
+class DebugData
 {
+public:
+	typedef typename std::shared_ptr<DebugData const> ConstPtr;
+
 	explicit DebugData(
-		langutil::SourceLocation _nativeLocation,
+		langutil::SourceLocation _nativeLocation = {},
 		langutil::SourceLocation _originLocation = {},
 		std::optional<int64_t> _astID = {}
 	):
-		nativeLocation(std::move(_nativeLocation)),
-		originLocation(std::move(_originLocation)),
-		astID(std::move(_astID))
+		m_nativeLocation(std::move(_nativeLocation)),
+		m_originLocation(std::move(_originLocation)),
+		m_astID(_astID)
 	{}
 
-	static std::shared_ptr<DebugData const> create(
-		langutil::SourceLocation _nativeLocation = {},
-		langutil::SourceLocation _originLocation = {},
-		std::optional<int64_t> _astID = {}
-	)
+	DebugData& nativeLocation(langutil::SourceLocation _nativeLocation)
 	{
-		return std::make_shared<DebugData const>(
-			std::move(_nativeLocation),
-			std::move(_originLocation),
-			std::move(_astID)
+		m_nativeLocation = _nativeLocation;
+		return *this;
+	}
+	[[nodiscard]] langutil::SourceLocation const& nativeLocation() const { return m_nativeLocation; }
+	[[nodiscard]] langutil::SourceLocation& nativeLocation() { return m_nativeLocation; }
+
+	DebugData& originLocation(langutil::SourceLocation _originLocation)
+	{
+		m_originLocation = _originLocation;
+		return *this;
+	}
+	[[nodiscard]] langutil::SourceLocation const& originLocation() const { return m_originLocation; }
+	[[nodiscard]] langutil::SourceLocation& originLocation() { return m_originLocation; }
+
+	DebugData& astID(int64_t _astID)
+	{
+		m_astID = _astID;
+		return *this;
+	}
+	[[nodiscard]] std::optional<int64_t> const& astID() const { return m_astID; }
+	[[nodiscard]] std::optional<int64_t>& astID() { return m_astID; }
+
+	[[nodiscard]] DebugData::ConstPtr constPtr() const
+	{
+		static DebugData::ConstPtr emptyDebugData{std::make_shared<DebugData const>(DebugData())};
+		if (nativeLocation() == emptyDebugData->nativeLocation() &&
+			originLocation() == emptyDebugData->originLocation() &&
+			astID() == emptyDebugData->astID())
+			return emptyDebugData;
+
+		static std::map<uint64_t , DebugData::ConstPtr> constPtrCache;
+		size_t hash = this->hash();
+		if (constPtrCache.find(hash) != constPtrCache.end())
+			return constPtrCache[hash];
+
+		DebugData::ConstPtr result = std::make_shared<DebugData const>(
+			m_nativeLocation,
+			m_originLocation,
+			m_astID
 		);
+		constPtrCache[hash] = result;
+		return result;
 	}
 
+	std::size_t hash() const
```

Comment:
Where is this used?

Reply:
Ah in the "constPtrCache"... well, I don't understand why you'd want that anyways :-).